das/daser.go

```diff
@@ -15,28 +15,152 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
-var log = logging.Logger("das")
+var (
+	log = logging.Logger("das")
 
-// TODO: parameters needs performance testing on real network to define optimal values
-const (
-	//  samplingRange is the maximum amount of headers processed in one job.
-	samplingRange = 100
+	ErrInvalidOptionValue = func(optionName string, value string) error {
+		errorMsg := fmt.Sprintf("das/daser: invalid option value: %s, cannot be %s", optionName, value)
+		return errors.New(errorMsg)
+	}
+)
+
+// optionSetter:
+//
+//	this is expected to be a closure that encloses the operation of parameter setting along side with
+//	the value to be set.
+//	example: both d and a are in the "global scope" relatively to the closure, thus no parameter passing is required.
+//	   func() {
+//
+// .       d.params.option = a
+//
+//	}
+type optionSetter func()
+type Option func(*DASer) error
+type OptionValidator func(value any, optionSetter optionSetter) error
+
+type parameters struct {
+	samplingRange     uint
+	concurrencyLimit  uint
+	bgStoreInterval   time.Duration
+	priorityQueueSize uint
+	genesisHeight     uint
+}
 
-	// concurrencyLimit defines the maximum amount of sampling workers running in parallel.
-	concurrencyLimit = 16
+func defaultParameters() parameters {
+	return parameters{
+		samplingRange:     100,
+		concurrencyLimit:  16,
+		bgStoreInterval:   10 * time.Minute,
+		priorityQueueSize: 16 * 4,
+		genesisHeight:     1,
+	}
+}
 
-	// backgroundStoreInterval is the period of time for background checkpointStore to perform a checkpoint backup.
-	backgroundStoreInterval = 10 * time.Minute
+// Option for: samplingRange
+func WithParamSamplingRange(samplingRange uint) Option {
```

Comment:
Why do we need to have `Param` prefixed to each option?

Reply:
Actually, we don't